# Importing Stuff

In [1]:
import warnings
warnings.filterwarnings("ignore")

import pandas as pd
import numpy as np

from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity
from sklearn.cluster import KMeans
from sklearn.preprocessing import StandardScaler, MultiLabelBinarizer, OneHotEncoder

from sentence_transformers import SentenceTransformer
from langdetect import detect, LangDetectException

import re
import nltk
from nltk.corpus import stopwords

2024-10-23 13:00:50.067799: I tensorflow/core/platform/cpu_feature_guard.cc:210] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: SSE4.1 SSE4.2 AVX AVX2 FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.


# Reading the cleaned dataset

In [2]:
df = pd.read_csv("data/dataset_cleaned.csv")
df.head()

Title                           Author  \
0              The Inner Game of Music  Barry Green, W. Timothy Gallwey   
1  The Nepalis a pictorial celebration  Deepak Thapa, Narendra. Pradhan   
2                        Lust for Life                     Irving Stone   
3                The Architecture Book                    Dk Publishing   
4        The World According to Colour                        James Fox   

    Price  List Price  Discount Amount  Limited Stock  Discount  \
0  1040.0      1040.0              0.0           True       0.0   
1  1500.0      1500.0              0.0           True       0.0   
2  1118.0      1118.0              0.0           True       0.0   
3  2238.0      2238.0              0.0           True       0.0   
4  1438.0      1438.0              0.0           True       0.0   

                                  Genre  Number of Pages  Weight  \
0                  Arts And Photography            248.0     200   
1                  Arts And Photography            116.0     515   
2  History Biography And Social Science            479.0     370   
3                  Arts And Photography            335.0    1200   
4  History Biography And Social Science            295.0     290   

            ISBN Language                                     Related Genres  \
0  9781447291725  English  ['Arts And Photography', 'Self Improvement And...   
1  9789993347378  English                 ['Arts And Photography', 'Nepali']   
2  9781787461390  English  ['History Biography And Social Science', 'Arts...   
3  9780241415030  English  ['Arts And Photography', 'Learning And Referen...   
4  9780141976655  English  ['History Biography And Social Science', 'Arts...   

                                 Subgenres  \
0     ['Music', 'Self Help', 'Psychology']   
1      ['Picture Books', 'Books on Nepal']   
2           ['Biography', 'Memoir', 'Art']   
3              ['Architecture', 'Science']   
4  ['History', 'Design', 'Art', 'Science']   

                                            Synopsis  \
0  The bestselling guide to improving musical per...   
1  Visitors to Nepal are very often struck by the...   
2  Since its initial publication in 1934, Irving ...   
3  Discover the key architectural concepts behind...   
4  A beguiling cultural history of colour, by one...   

                                                 URL  
0  https://booksmandala.com/books/the-inner-game-...  
1  https://booksmandala.com/books/the-nepalis-a-p...  
2  https://booksmandala.com/books/lust-for-life-3...  
3  https://booksmandala.com/books/the-architectur...  
4  https://booksmandala.com/books/the-world-accor...

# Preprocess

## Get rid of missing rows

In [3]:
na = df.isna().sum()
na[na > 0]

Author         3
Subgenres    136
dtype: int64

In [4]:
df.dropna(subset=["Author"], inplace=True)

## Remove "Unassigned" Authors

In [5]:
df["Author"].value_counts()

Author
Unassigned                      27
Jeff Kinney                     24
Hergé                           23
Thich Nhat Hanh                 16
Kentaro Miura                   16
                                ..
Luo Di Cheng Qiu                 1
Jim Starlin                      1
Negi Haruba                      1
Syougo Kinugasa, Yuyu Ichino     1
Peter Matthiessen                1
Name: count, Length: 1347, dtype: int64

In [6]:
df = df[df["Author"] != "Unassigned"]

## Remove Redundant Synopses

In [7]:
df["Language"].value_counts()

Language
English     1774
Nepali       116
French        13
Hindi         13
German         3
Japanese       1
Bengali        1
Name: count, dtype: int64

In [8]:
df[df["Language"] != "English"]["Synopsis"]

109     Timeless lessons on wealth, greed, and happine...
258     " Je me représente très bien l'époque où les r...
259     फ़ैज़ अहमद फै़ज़ उर्दू के बहुत ही जाने-माने कव...
260     ऋषि अष्टावक्र और राजा जनक के संवाद का संकलन है...
262     本書ではネパール・ヒマラヤの中でも、交通の便、現地事情、内外トレッキング・エージェントの実績...
                              ...                        
1191    त्यो मेरो अनुभूतिभन्दा बाहिरको मूल्य हो आदिमात...
1192    Sumnima a famous Nepali novel by B P Koirala, ...
1321         Hindi verse translation of the Bhagavadgītā.
1388    धर्मलाई बुझ्ने मानिस दुर्लभ हुन्छ, बुझेर अरूला...
1639    This collection of new translations brings tog...
Name: Synopsis, Length: 147, dtype: object

In [9]:
df.duplicated(subset=["ISBN", "Synopsis"]).sum()

0

In [10]:
df[df["Language"] != "English"]["Synopsis"]

109     Timeless lessons on wealth, greed, and happine...
258     " Je me représente très bien l'époque où les r...
259     फ़ैज़ अहमद फै़ज़ उर्दू के बहुत ही जाने-माने कव...
260     ऋषि अष्टावक्र और राजा जनक के संवाद का संकलन है...
262     本書ではネパール・ヒマラヤの中でも、交通の便、現地事情、内外トレッキング・エージェントの実績...
                              ...                        
1191    त्यो मेरो अनुभूतिभन्दा बाहिरको मूल्य हो आदिमात...
1192    Sumnima a famous Nepali novel by B P Koirala, ...
1321         Hindi verse translation of the Bhagavadgītā.
1388    धर्मलाई बुझ्ने मानिस दुर्लभ हुन्छ, बुझेर अरूला...
1639    This collection of new translations brings tog...
Name: Synopsis, Length: 147, dtype: object

In [11]:
detect(df['Synopsis'][291])

'nl'

In [12]:
df.iloc[291]

Title                                    Flickan som lekte med elden
Author                                                 Stieg Larsson
Price                                                          400.0
List Price                                                     400.0
Discount Amount                                                  0.0
Limited Stock                                                   True
Discount                                                         0.0
Genre                                              Foreign Languages
Number of Pages                                                260.0
Weight                                                           250
ISBN                                                   9789170014833
Language                                                     English
Related Genres                                 ['Foreign Languages']
Subgenres                                                 ['German']
Synopsis           On the eve of p

In [13]:
def detect_language(row):
    try:
        return detect(row['Synopsis']) 
    except LangDetectException:
        return row['Language']

df['Language'] = df.apply(detect_language, axis=1)

In [14]:
df['Language'].unique()

array(['en', 'fr', 'hi', 'ja', 'nl', 'de', 'ne', 'English', 'cy', 'lv',
       'no'], dtype=object)

In [15]:
df[df['Language'] == 'English']

Title      Author  Price  List Price  \
413  George's Marvellous Medicine  Roald Dahl  560.0       560.0   

     Discount Amount  Limited Stock  Discount           Genre  \
413              0.0          False       0.0  Kids And Teens   

     Number of Pages  Weight           ISBN Language  \
413            136.0     200  9780241558485  English   

                                   Related Genres  \
413  ['Kids And Teens', 'Fiction And Literature']   

                                             Subgenres Synopsis  \
413  ['Childrens', 'Fantasy', 'Classics', 'Young Ad...      560   

                                                   URL  
413  https://booksmandala.com/books/georges-marvell...

In [16]:
import requests
from dotenv import load_dotenv
import os

load_dotenv()

def get_synopsis(isbn: str) -> str | None:
    # GET YOUR OWN API KEY!!!
    response = requests.get(f"https://www.googleapis.com/books/v1/volumes?q=isbn:{isbn}&key={os.getenv('API_KEY')}")
    data = response.json()

    if "items" not in data or len(data["items"]) == 0:
        return None
        
    new_response = requests.get(data["items"][0].get("selfLink", {}))
    new_data = new_response.json()
    new_data = new_data.get("volumeInfo", {})
    return new_data.get("description", None)

df["Synopsis"] = df.apply(lambda row: row["Synopsis"] if row["Language"] != "English" else get_synopsis(row["ISBN"]), axis=1)

In [17]:
df["Language"] = df.apply(lambda row: row["Language"] if row["Language"] != "English" else "en", axis=1)

In [18]:
df["Synopsis"] = df["Synopsis"].str.replace(r"<.*?>", '', regex=True)

## Safe parse genres and subgenres

In [19]:
import ast

def safe_eval_list(x: str | list[str]) -> list:
    if pd.isna(x):
        return []

    if isinstance(x, list):
        return x

    try:
        evaluated = ast.literal_eval(x)
        if isinstance(evaluated, list):
            return evaluated
    except:
        return []

In [20]:
df["Related Genres"] = df["Related Genres"].apply(safe_eval_list)

In [21]:
df["Subgenres"] = df["Subgenres"].apply(safe_eval_list)

In [22]:
df[["Related Genres", "Subgenres"]]

Related Genres  \
0     [Arts And Photography, Self Improvement And Re...   
1                        [Arts And Photography, Nepali]   
2     [History Biography And Social Science, Arts An...   
3        [Arts And Photography, Learning And Reference]   
4     [History Biography And Social Science, Arts An...   
...                                                 ...   
1946             [History Biography And Social Science]   
1947                                   [Travel, Nepali]   
1948                           [Fiction And Literature]   
1949  [History Biography And Social Science, Busines...   
1950             [History Biography And Social Science]   

                                          Subgenres  
0                    [Music, Self Help, Psychology]  
1                   [Picture Books, Books on Nepal]  
2                          [Biography, Memoir, Art]  
3                           [Architecture, Science]  
4                   [History, Design, Art, Science]  
...                                             ...  
1946                                       [Memoir]  
1947  [Climbing and Mountaineering, Books on Nepal]  
1948                       [Classics, Contemporary]  
1949                  [Memoir, Biography, Business]  
1950                                [Autobiography]  

[1921 rows x 2 columns]

In [23]:
df.dropna(subset=["Synopsis"], inplace=True)

## **Related Genres** and **Subgenres**

In [24]:
mlb_genres = MultiLabelBinarizer()

genres_encoded = pd.DataFrame(mlb_genres.fit_transform(df["Related Genres"]), 
                              columns=mlb_genres.classes_, 
                              index=df.index)

In [25]:
genres_encoded

Arts And Photography  Business And Investing  Fiction And Literature  \
0                        1                       0                       0   
1                        1                       0                       0   
2                        1                       0                       0   
3                        1                       0                       0   
4                        1                       0                       0   
...                    ...                     ...                     ...   
1946                     0                       0                       0   
1947                     0                       0                       0   
1948                     0                       0                       1   
1949                     0                       1                       0   
1950                     0                       0                       0   

      Foreign Languages  History Biography And Social Science  Kids And Teens  \
0                     0                                     1               0   
1                     0                                     0               0   
2                     0                                     1               0   
3                     0                                     0               0   
4                     0                                     1               0   
...                 ...                                   ...             ...   
1946                  0                                     1               0   
1947                  0                                     0               0   
1948                  0                                     0               0   
1949                  0                                     1               0   
1950                  0                                     1               0   

      Learning And Reference  Lifestyle And Wellness  \
0                          0                       0   
1                          0                       0   
2                          0                       0   
3                          1                       0   
4                          1                       0   
...                      ...                     ...   
1946                       0                       0   
1947                       0                       0   
1948                       0                       0   
1949                       0                       0   
1950                       0                       0   

      Manga And Graphic Novels  Miscellaneous  Nature  Nepali  \
0                            0              0       0       0   
1                            0              0       0       1   
2                            0              0       0       0   
3                            0              0       0       0   
4                            0              0       0       0   
...                        ...            ...     ...     ...   
1946                         0              0       0       0   
1947                         0              0       0       1   
1948                         0              0       0       0   
1949                         0              0       0       0   
1950                         0              0       0       0   

      Rare Coffee Table Books  Religion  Self Improvement And Relationships  \
0                           0         0                                   1   
1                           0         0                                   0   
2                           0         0                                   0   
3                           0         0                                   0   
4                           0         0                                   0   
...                       ...       ...                                 ...   
1946                        0         0                                   0   
1947      

In [26]:
mlb_subgenres = MultiLabelBinarizer()

subgenres_encoded = pd.DataFrame(mlb_subgenres.fit_transform(df["Subgenres"]), 
                              columns=mlb_subgenres.classes_, 
                              index=df.index)

In [27]:
print(len(subgenres_encoded.columns))
subgenres_encoded.columns

139


Index(['Action and Adventure', 'Adult Fiction', 'Ages 3 to 5', 'Ages 6 to 8',
       'Ages 9 to 12', 'Animals and Pets', 'Anthology', 'Anthropology',
       'Architecture', 'Art',
       ...
       'Stress Management', 'Tarot', 'Teens and Young Adult',
       'Thriller and Suspense', 'Time Management', 'Travel Guide Books',
       'Trees and Plants', 'True Crime', 'Womens Fiction', 'Young Adult'],
      dtype='object', length=139)

One hot encoding all related genres and subgenres will result in $18 + 139 = 157$ additional columns, increasing the dimensionality.

## Encoding

In [28]:
df['genres_str'] = df["Related Genres"].apply(lambda x: ','.join(x))
df['subgenres_str'] = df['Subgenres'].apply(lambda x: ','.join(x))

In [29]:
df = df[["Title", "Author", "Genre", "Language", "Synopsis", "genres_str", "subgenres_str", "ISBN", "URL"]]

In [30]:
df["combined_genres"] = df["genres_str"] + "," + df["subgenres_str"]

In [31]:
model = SentenceTransformer('paraphrase-multilingual-MiniLM-L12-v2')

In [32]:
df.reset_index(inplace=True)
df["synopsis_embeddings"] = list(model.encode(df["Synopsis"], show_progress_bar=True))

Batches:   0%|          | 0/61 [00:00<?, ?it/s]

### Treating Genres and Subgenres as Combined Features

In [33]:
tfidf = TfidfVectorizer(stop_words='english')
genre_vectors = tfidf.fit_transform(df["combined_genres"])

### Combine and Add Weights

In [34]:


weights = {
    "genre_weight": 3.0,
    "synopsis_weight": 2.0
}

synopsis_embeddings_matrix = np.vstack(df["synopsis_embeddings"].values)

combined_features = hstack([
    weights["genre_weight"] * genre_vectors,
    weights["synopsis_weight"] * synopsis_embeddings_matrix
])

# Calculate Similarity

## Cosine Similarity

In [35]:
similarity_matrix = cosine_similarity(combined_features)
similarity_df = pd.DataFrame(similarity_matrix, index=df['Title'], columns=df['Title'])

In [36]:
def get_recommendations(similarity_df: pd.DataFrame,
                        df: pd.DataFrame,
                        title: str, 
                        n: int = 5, 
                        columns: list[str] = ["Title", "Author"]
                       ) -> pd.DataFrame:
        idx = similarity_df.index.get_loc(title)
        sim_scores = list(enumerate(similarity_df.iloc[idx]))
    
        # sort the books based on the similarity scores
        sim_scores = sorted(sim_scores, key=lambda x: x[1], reverse=True)
    
        sim_scores = sim_scores[1:n+1]  # exclude the first one as it's the book itself
    
        # get the book indices
        book_indices = [i[0] for i in sim_scores]
    
        # top n most similar books
        return df[columns].iloc[book_indices]

In [37]:
get_recommendations(similarity_df, df, "Peter Pan", n=10, columns=["Title", "Author", "combined_genres", "Language", "URL"])

Title  \
346                                          Pater Pan   
438                                    The Jungle Book   
50                          The 7 Habits of Happy Kids   
8    Colouring book : Copy Colour Fruits and Vegeta...   
37   Colouring book : Copy Colouring Book Noisy Din...   
446        Gruffalo, Where Are You?: a Felt Flaps Book   
688                                   Destination Moon   
665                              Explorers on the Moon   
357  Brain Activity Book for Kids - 200+ Activities...   
657                                The Calculus Affair   

                                               Author  \
346                                        Fun Studio   
438                                        Miriam Bos   
50                                         Sean Covey   
8                              Om Books International   
37                             Om Books International   
446                                    Julia Donaldon   
688                                             Hergé   
665  Hergé, Leslie Lonsdale-Cooper, Michael R. Turner   
357                                       Maple Press   
657                                             Hergé   

                                       combined_genres Language  \
346               Kids And Teens,Baby to 2,Ages 3 to 5       en   
438  Arts And Photography,Kids And Teens,Picture Bo...       en   
50   Arts And Photography,Kids And Teens,Picture Bo...       en   
8    Kids And Teens,Arts And Photography,Ages 3 to ...       en   
37   Kids And Teens,Arts And Photography,Ages 3 to ...       en   
446  Arts And Photography,Kids And Teens,Picture Bo...       en   
688  Kids And Teens,Manga And Graphic Novels,Ages 9...       en   
665  Kids And Teens,Manga And Graphic Novels,Ages 9...       en   
357  Kids And Teens,Ages 3 to 5,Ages 6 to 8,Childre...       en   
657  Kids And Teens,Manga And Graphic Novels,Ages 9...       en   

                                                   URL  
346     https://booksmandala.com/books/pater-pan-49862  
438  https://booksmandala.com/books/the-jungle-book...  
50   https://booksmandala.com/books/the-7-habits-of...  
8    https://booksmandala.com/books/colouring-book-...  
37   https://booksmandala.com/books/colouring-book-...  
446  https://booksmandala.com/books/gruffalo-where-...  
688  https://booksmandala.com/books/destination-moo...  
665  https://booksmandala.com/books/explorers-on-th...  
357  https://booksmandala.com/books/brain-activity-...  
657  https://booksmandala.com/books/the-calculus-af...

In [38]:
get_recommendations(similarity_df, df, "Diary of a Wimpy Kid", n=10, columns=["Title", "Author", "combined_genres", "Language", "URL"])

Title       Author  \
341              Diary Of A Wimpy Kid ; The Ugly Truth  Jeff Kinney   
362                Diary of a Wimpy Kid: Wrecking Ball  Jeff Kinney   
355                                        Cabin Fever  Jeff Kinney   
787                   Diary Of A Wimpy Kid: No Brainer  Jeff Kinney   
741  Diary of an Awesome Friendly Kid: Rowley Jeffe...  Jeff Kinney   
359                                      Rodrick Rules  Jeff Kinney   
371                Diary of a Wimpy Kid 10. Old School  Jeff Kinney   
434                   Diary Of A Wimpy Kid ; Hard Luck  Jeff Kinney   
377                          The Wimpy Kid Movie Diary  Jeff Kinney   
427     Diary of a Wimpy Kid: Diper Overlode (Book 17)  Jeff Kinney   

                                       combined_genres Language  \
341                        Kids And Teens,Ages 9 to 12       en   
362                        Kids And Teens,Ages 9 to 12       en   
355                        Kids And Teens,Ages 9 to 12       en   
787  Kids And Teens,Fiction And Literature,Manga An...       en   
741  Fiction And Literature,Kids And Teens,Manga An...       en   
359            Kids And Teens,Ages 6 to 8,Ages 9 to 12       en   
371  Fiction And Literature,Kids And Teens,Young Ad...       en   
434                        Kids And Teens,Ages 9 to 12       en   
377                        Kids And Teens,Ages 9 to 12       en   
427                        Kids And Teens,Ages 9 to 12       en   

                                                   URL  
341  https://booksmandala.com/books/diary-of-a-wimp...  
362  https://booksmandala.com/books/diary-of-a-wimp...  
355   https://booksmandala.com/books/cabin-fever-37914  
787  https://booksmandala.com/books/diary-of-a-wimp...  
741  https://booksmandala.com/books/diary-of-an-awe...  
359  https://booksmandala.com/books/rodrick-rules-3...  
371  https://booksmandala.com/books/diary-of-a-wimp...  
434  https://booksmandala.com/books/diary-of-a-wimp...  
377  https://booksmandala.com/books/the-wimpy-kid-m...  
427  https://booksmandala.com/books/diary-of-a-wimp...

In [39]:
get_recommendations(similarity_df, df, "Satori in Paris", n=5, columns=["Title", "Author", "combined_genres", "Language", "URL"])

Title                      Author  \
1878                Lonesome Traveler                Jack Kerouac   
1610  Metamorphosis and Other Stories                 Franz Kafka   
1898                      On the Road  Jack Kerouac, Ann Charters   
1561  Metamorphosis and Other Stories                 Franz Kafka   
226                    Norwegian Wood             Haruki Murakami   

                                        combined_genres Language  \
1878                    Fiction And Literature,Classics       en   
1610  Fiction And Literature,Spirituality And Philos...       en   
1898                    Fiction And Literature,Classics       en   
1561  Fiction And Literature,Spirituality And Philos...       en   
226         Fiction And Literature,Contemporary,Romance       en   

                                                    URL  
1878  https://booksmandala.com/books/lonesome-travel...  
1610  https://booksmandala.com/books/metamorphosis-a...  
1898   https://booksmandala.com/books/on-the-road-49631  
1561  https://booksmandala.com/books/metamorphosis-a...  
226   https://booksmandala.com/books/norwegian-wood-...

## K-Nearest Neighbors

In [41]:
# scaling the features

# scaling the synopsis (dense)
scaler = StandardScaler()
synopsis_embeddings_scaled = scaler.fit_transform(synopsis_embeddings_matrix)

combined_features_scaled = np.hstack([
    synopsis_embeddings_scaled,
    genre_vectors.toarray()
])

In [45]:


k = 10 # number of recommendations (neighbors)

# KNN using cosine distance
knn = NearestNeighbors(n_neighbors=k, metric='cosine', algorithm='brute')
knn.fit(combined_features_scaled)

NearestNeighbors(algorithm='brute', metric='cosine', n_neighbors=10)

In [52]:
def recommend_knn(book_title, 
                  df, 
                  knn_model, 
                  combined_features_scaled, 
                  n_recommendations=5,
                  columns=['Title', 'Author', 'combined_genres', 'Distance']):
    try:
        book_idx = df[df['Title'] == book_title].index[0]
    except IndexError:
        return f'Book "{book_title}" not found.'

    # Get the feature vector for the book
    book_vector = combined_features_scaled[book_idx].reshape(1, -1)
    
    # Find k nearest neighbors (including the book itself)
    distances, indices = knn_model.kneighbors(book_vector, n_neighbors=n_recommendations+1)
    
    # Get indices of the recommended books (skipping the first as it's the book itself)
    recommended_indices = indices[0][1:]
    recommended_distances = distances[0][1:]
    
    # Create a DataFrame with the recommended books and distances
    recommendations_df = df.iloc[recommended_indices].copy()  # Copy to avoid SettingWithCopyWarning
    recommendations_df['Distance'] = recommended_distances  # Add distances to the DataFrame
    
    return recommendations_df[columns]

In [53]:
recommend_knn("Satori in Paris", df, knn, combined_features_scaled)

Title         Author  \
1878                Lonesome Traveler   Jack Kerouac   
1610  Metamorphosis and Other Stories    Franz Kafka   
1561  Metamorphosis and Other Stories    Franz Kafka   
1501               Dharmayoddha Kalki   Kevin Missal   
1158                         Jay Vudi  Bhairav Aryal   

                                        combined_genres  Distance  
1878                    Fiction And Literature,Classics  0.542328  
1610  Fiction And Literature,Spirituality And Philos...  0.597776  
1561  Fiction And Literature,Spirituality And Philos...  0.598379  
1501  Fiction And Literature,Spirituality And Philos...  0.655412  
1158                           Nepali,Nepali Literature  0.657841

In [54]:
recommend_knn("Diary of a Wimpy Kid", df, knn, combined_features_scaled)

Title       Author  \
341              Diary Of A Wimpy Kid ; The Ugly Truth  Jeff Kinney   
362                Diary of a Wimpy Kid: Wrecking Ball  Jeff Kinney   
787                   Diary Of A Wimpy Kid: No Brainer  Jeff Kinney   
355                                        Cabin Fever  Jeff Kinney   
741  Diary of an Awesome Friendly Kid: Rowley Jeffe...  Jeff Kinney   

                                       combined_genres  Distance  
341                        Kids And Teens,Ages 9 to 12  0.387765  
362                        Kids And Teens,Ages 9 to 12  0.416114  
787  Kids And Teens,Fiction And Literature,Manga An...  0.417212  
355                        Kids And Teens,Ages 9 to 12  0.442148  
741  Fiction And Literature,Kids And Teens,Manga An...  0.463581

In [55]:
recommend_knn("Jay Vudi", df, knn, combined_features_scaled)

Title                                 Author  \
527   Ratna's basic Nepali dictionary  Shyam P. Wagley, Bijay Kumar Rauniyar   
1099                        Ghatmandu                        Kumar Nagarkoti   
196        Arresting god in kathmandu                        Samrat Upadhyay   
1146                      Damini Bhir                         Rajan Mukarung   
1156                      Ijajatpatra                          Sarthak Karki   

                                      combined_genres  Distance  
527               Learning And Reference,Dictionaries  0.542067  
1099         Nepali,Nepali Literature,Nepali Language  0.554654  
196   Fiction And Literature,Short Story,Contemporary  0.575301  
1146         Nepali,Nepali Literature,Nepali Language  0.576937  
1156                         Nepali,Nepali Literature  0.580010

In [56]:
recommend_knn("Big Magic", df, knn, combined_features_scaled, n_recommendations=10)

Title  \
1835                                      Eat Pray Love   
944                         The Signature of All Things   
457                               How Emotions Are Made   
495                                    The Artist's Way   
319                                           Jane Eyre   
313                           The Diary of a young girl   
229   I Don't Love You Anymore: Moving On and Living...   
1427  The One Thing: The Suprisingly Simple Truth Be...   
1451                       Everything I Know about Love   
26                                   Art of Mindfulness   

                        Author  \
1835         Elizabeth Gilbert   
944          Elizabeth Gilbert   
457       Lisa Feldman Barrett   
495              Julia Cameron   
319           Charlotte Brontë   
313                 Anne Frank   
229              Rithvik Singh   
1427  Gary Keller, Jay Papasan   
1451            Dolly Alderton   
26              Emma Farrarons   

                                        combined_genres  Distance  
1835  History Biography And Social Science,Fiction A...  0.205838  
944           Fiction And Literature,Historical Fiction  0.451176  
457   History Biography And Social Science,Learning ...  0.456232  
495   Arts And Photography,Self Improvement And Rela...  0.502150  
319   Fiction And Literature,History Biography And S...  0.518982  
313   History Biography And Social Science,Fiction A...  0.623711  
229             Fiction And Literature,Poetry and Prose  0.624618  
1427  Business And Investing,Self Improvement And Re...  0.627458  
1451  Self Improvement And Relationships,History Bio...  0.639673  
26    Arts And Photography,Self Improvement And Rela...  0.647875